File: adws-bun/src/workflows/adw-plan-implement-update-teamwork-task.ts
#!/usr/bin/env bun
/**
 * Plan-Implement workflow with Teamwork task updates.
 *
 * Complex workflow: /plan → /implement → /update_teamwork_task
 * Supports prototype-specific planning commands.
 *
 * TypeScript/Bun port of adws/adw_plan_implement_update_teamwork_task.py
 */

import { join } from 'path';
import { executeTemplate } from '@/modules/agent';
import type { AgentPromptResponse } from '@/modules/data-models';
import { setupLogger } from '@/modules/utils';

/**
 * Map prototype type to corresponding plan command.
 */
function getPlanCommand(prototype?: string): string {
  const prototypeMap: Record<string, string> = {
    'vite_vue': '/plan_vite_vue',
    'uv_script': '/plan_uv_script',
    'bun_scripts': '/plan_bun_scripts',
    'uv_mcp': '/plan_uv_mcp',
  };

  return prototypeMap[prototype || ''] || '/plan';
}

/**
 * Find the most recently created plan file in specs directory.
 */
async function findLatestPlanFile(specsDir: string): Promise<string | null> {
  try {
    const files: string[] = [];
    const glob = new Bun.Glob('plan-*.md');

    for await (const file of glob.scan({ cwd: specsDir })) {
      files.push(file);
    }

    if (files.length === 0) {
      return null;
    }

    // Get file stats and sort by modified time
    const filesWithStats = await Promise.all(
      files.map(async (file) => {
        const path = join(specsDir, file);
        const stat = await Bun.file(path).stat();
        return { path, mtime: stat.mtime };
      })
    );

    // Sort by mtime descending (most recent first)
    filesWithStats.sort((a, b) => b.mtime.getTime() - a.mtime.getTime());

    return filesWithStats[0]?.path || null;
  } catch (error) {
    return null;
  }
}

/**
 * Get current git commit hash.
 */
async function getCurrentCommitHash(workingDir: string): Promise<string | null> {
  try {
    const proc = Bun.spawn({
      cmd: ['git', 'rev-parse', 'HEAD'],
      cwd: workingDir,
      stdout: 'pipe',
      stderr: 'pipe',
    });

    const exitCode = await proc.exited;
    if (exitCode === 0) {
      const stdout = await new Response(proc.stdout).text();
      return stdout.trim().slice(0, 8); // First 8 characters
    }

    return null;
  } catch (error) {
    return null;
  }
}

/**
 * Main workflow execution function.
 */
async function main(
  adwId: string,
  taskId: string,
  taskDescription: string,
  worktreeName: string,
  prototype: string = '',
  model: 'sonnet' | 'opus' = 'sonnet',
  _projectId?: string
): Promise<number> {
  // Setup logger
  const logger = await setupLogger(adwId, 'plan-implement-update-teamwork');

  logger.info('=== Starting Plan-Implement Workflow ===');
  logger.info(`ADW ID: ${adwId}`);
  logger.info(`Task ID: ${taskId}`);
  logger.info(`Worktree: ${worktreeName}`);
  logger.info(`Prototype: ${prototype || 'None'}`);
  logger.info(`Model: ${model}`);

  let workflowSuccess = false;
  let commitHash: string | null = null;
  let errorMessage: string | null = null;
  let planPath: string | null = null;
  let implementResponse: AgentPromptResponse | null = null;

  // Determine working directory
  const worktreePath = join(process.cwd(), '..', 'trees', worktreeName, 'tac8_app4__agentic_prototyping');
  const workingDir = await Bun.file(worktreePath).exists()
    ? worktreePath
    : process.cwd();

  logger.info(`Working directory: ${workingDir}`);

  try {
    // === Phase 1: Planning ===
    logger.info('=== Phase 1: Planning ===');

    const planCommand = getPlanCommand(prototype);
    logger.info(`Using planning command: ${planCommand}`);

    const planResponse = await executeTemplate({
      agentName: `plan-${worktreeName}`,
      slashCommand: planCommand,
      args: [adwId, taskDescription],
      adwId,
      model,
      workingDir,
    });

    if (!planResponse.success) {
      errorMessage = `Planning failed: ${planResponse.output}`;
      logger.error(errorMessage);
      throw new Error(errorMessage);
    }

    logger.info('Planning completed successfully');

    // Find the generated plan file
    const specsDir = join(workingDir, 'specs');
    if (await Bun.file(specsDir).exists()) {
      planPath = await findLatestPlanFile(specsDir);
      if (planPath) {
        logger.info(`Found plan file: ${planPath}`);
      } else {
        logger.warn('No plan file found in specs/');
      }
    } else {
      logger.warn('specs/ directory not found');
    }

    // === Phase 2: Implementation ===
    logger.info('=== Phase 2: Implementation ===');

    const implementArgs = planPath ? [adwId, planPath] : [adwId, taskDescription];

    implementResponse = await executeTemplate({
      agentName: `implement-${worktreeName}`,
      slashCommand: '/implement',
      args: implementArgs,
      adwId,
      model,
      workingDir,
    });

    if (!implementResponse.success) {
      errorMessage = `Implementation failed: ${implementResponse.output}`;
      logger.error(errorMessage);
      throw new Error(errorMessage);
    }

    logger.info('Implementation completed successfully');
    workflowSuccess = true;

    // Try to get commit hash
    commitHash = await getCurrentCommitHash(workingDir);
    if (commitHash) {
      logger.info(`Commit hash: ${commitHash}`);
    }
  } catch (error) {
    errorMessage = `Exception during workflow: ${error}`;
    logger.error(errorMessage);
  }

  // === Phase 3: Update Teamwork ===
  logger.info('=== Phase 3: Updating Teamwork ===');

  // Determine final status
  const updateStatus = workflowSuccess && commitHash ? 'Done' : 'Failed';

  // Build result payload
  const updateContent = {
    status: updateStatus,
    adw_id: adwId,
    commit_hash: commitHash || '',
    error: errorMessage || '',
    timestamp: new Date().toISOString(),
    model,
    workflow: 'plan-implement-update',
    worktree_name: worktreeName,
    prototype: prototype || '',
    plan_path: planPath || '',
    result: implementResponse?.output || '',
  };

  // Execute update
  logger.info(`Updating Teamwork task to ${updateStatus}...`);

  const updateResponse = await executeTemplate({
    agentName: `teamwork-updater-${worktreeName}`,
    slashCommand: '/update_teamwork_task',
    args: [taskId, updateStatus, JSON.stringify(updateContent)],
    adwId,
    model,
    workingDir,
  });

  if (updateResponse.success) {
    logger.info(`Successfully updated Teamwork task ${taskId} to ${updateStatus}`);
  } else {
    logger.error(`Failed to update Teamwork task: ${updateResponse.output}`);
  }

  logger.info('=== Plan-Implement Workflow Complete ===');

  return workflowSuccess ? 0 : 1;
}

// Parse command line arguments
const args = process.argv.slice(2);

if (args.length < 4) {
  console.error('Usage: adw-plan-implement-update-teamwork-task.ts <adw_id> <task_id> <task_description> <worktree_name> [prototype] [--model <model>] [--project-id <project_id>]');
  process.exit(1);
}

const adwId = args[0] as string;
const taskId = args[1] as string;
const taskDescription = args[2] as string;
const worktreeName = args[3] as string;
let prototype = args[4] || '';
let model: 'sonnet' | 'opus' = 'sonnet';
let projectId: string | undefined;

// Parse optional flags
for (let i = 5; i < args.length; i++) {
  const arg = args[i];
  if (arg === '--model' && i + 1 < args.length) {
    const modelArg = args[i + 1];
    if (modelArg === 'opus' || modelArg === 'sonnet') {
      model = modelArg;
    }
    i++;
  } else if (arg === '--project-id' && i + 1 < args.length) {
    projectId = args[i + 1];
    i++;
  }
}

// Run the workflow
main(adwId, taskId, taskDescription, worktreeName, prototype, model, projectId)
  .then((exitCode) => process.exit(exitCode))
  .catch((error) => {
    console.error(`Fatal error: ${error}`);
    process.exit(2);
  });

File: archive/python-adws-20251111/adws/adw_plan_implement_update_teamwork_task.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.10"
# dependencies = [
#   "pydantic",
#   "python-dotenv",
#   "click",
#   "rich",
# ]
# ///
"""
Plan-Implement workflow with Teamwork task updates.

Complex workflow: /plan → /implement → /update_teamwork_task
Supports prototype-specific planning commands.
"""

import os
import sys
import json
import logging
import argparse
from pathlib import Path
from datetime import datetime
from typing import Optional

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent))

from adw_modules.agent import execute_template, AgentTemplateRequest
from adw_modules.data_models import TeamworkTaskUpdate

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


def main(
    adw_id: str,
    task_id: str,
    task_description: str,
    worktree_name: str,
    prototype: str = "",
    model: str = "sonnet",
    project_id: Optional[str] = None
):
    """Execute plan-implement workflow and update Teamwork task."""

    logger.info("=== Starting Plan-Implement Workflow ===")
    logger.info(f"ADW ID: {adw_id}")
    logger.info(f"Task ID: {task_id}")
    logger.info(f"Worktree: {worktree_name}")
    logger.info(f"Prototype: {prototype or 'None'}")
    logger.info(f"Model: {model}")

    workflow_success = False
    commit_hash = None
    error_message = None
    plan_path = None

    # Determine working directory
    worktree_path = Path(f"../trees/{worktree_name}/tac8_app4__agentic_prototyping")

    if worktree_path.exists():
        working_dir = str(worktree_path.absolute())
        logger.info(f"Using existing worktree: {working_dir}")
    else:
        working_dir = os.getcwd()
        logger.info(f"Worktree not found, using current directory: {working_dir}")

    try:
        # === Phase 1: Planning ===
        logger.info("=== Phase 1: Planning ===")

        # Determine plan command based on prototype type
        if prototype:
            if prototype == "uv_script":
                plan_command = "/plan_uv_script"
            elif prototype == "vite_vue":
                plan_command = "/plan_vite_vue"
            elif prototype == "bun_scripts":
                plan_command = "/plan_bun_scripts"
            elif prototype == "uv_mcp":
                plan_command = "/plan_uv_mcp"
            else:
                logger.warning(f"Unknown prototype type: {prototype}, using /plan")
                plan_command = "/plan"
        else:
            plan_command = "/plan"

        logger.info(f"Using planning command: {plan_command}")

        plan_request = AgentTemplateRequest(
            agent_name=f"plan-{worktree_name}",
            slash_command=plan_command,
            args=[adw_id, task_description],
            adw_id=adw_id,
            model=model,
            working_dir=working_dir
        )

        plan_response = execute_template(plan_request)

        if not plan_response.success:
            error_message = f"Planning failed: {plan_response.output}"
            logger.error(error_message)
            raise Exception(error_message)

        logger.info("Planning completed successfully")

        # Try to find the generated plan file
        specs_dir = Path(working_dir) / "specs"
        if specs_dir.exists():
            # Look for most recently created plan file
            plan_files = list(specs_dir.glob("plan-*.md"))
            if plan_files:
                plan_path = str(max(plan_files, key=lambda p: p.stat().st_mtime))
                logger.info(f"Found plan file: {plan_path}")
            else:
                logger.warning("No plan file found in specs/")
        else:
            logger.warning("specs/ directory not found")

        # === Phase 2: Implementation ===
        logger.info("=== Phase 2: Implementation ===")

        if plan_path:
            implement_args = [adw_id, plan_path]
        else:
            # Fallback: use task description
            implement_args = [adw_id, task_description]

        implement_request = AgentTemplateRequest(
            agent_name=f"implement-{worktree_name}",
            slash_command="/implement",
            args=implement_args,
            adw_id=adw_id,
            model=model,
            working_dir=working_dir
        )

        implement_response = execute_template(implement_request)

        if not implement_response.success:
            error_message = f"Implementation failed: {implement_response.output}"
            logger.error(error_message)
            raise Exception(error_message)

        logger.info("Implementation completed successfully")
        workflow_success = True

        # Try to get commit hash from git
        try:
            import subprocess
            result = subprocess.run(
                ["git", "rev-parse", "HEAD"],
                cwd=working_dir,
                capture_output=True,
                text=True,
                timeout=10
            )
            if result.returncode == 0:
                commit_hash = result.stdout.strip()[:8]
                logger.info(f"Commit hash: {commit_hash}")
        except Exception as e:
            logger.warning(f"Could not get commit hash: {e}")

    except Exception as e:
        error_message = f"Exception during workflow: {str(e)}"
        logger.error(error_message)

    # === Phase 3: Update Teamwork ===
    logger.info("=== Phase 3: Updating Teamwork ===")

    # Determine final status
    update_status = "Done" if workflow_success and commit_hash else "Failed"

    # Build result payload
    update_content = {
        "status": update_status,
        "adw_id": adw_id,
        "commit_hash": commit_hash or "",
        "error": error_message or "",
        "timestamp": datetime.now().isoformat(),
        "model": model,
        "workflow": "plan-implement-update",
        "worktree_name": worktree_name,
        "prototype": prototype or "",
        "plan_path": plan_path or "",
        "result": implement_response.output if workflow_success and implement_response else ""
    }

    # Execute update
    logger.info(f"Updating Teamwork task to {update_status}...")

    update_request = AgentTemplateRequest(
        agent_name=f"teamwork-updater-{worktree_name}",
        slash_command="/update_teamwork_task",
        args=[task_id, update_status, json.dumps(update_content)],
        adw_id=adw_id,
        model=model,
        working_dir=working_dir
    )

    update_response = execute_template(update_request)

    if update_response.success:
        logger.info(f"Successfully updated Teamwork task {task_id} to {update_status}")
    else:
        logger.error(f"Failed to update Teamwork task: {update_response.output}")

    logger.info("=== Plan-Implement Workflow Complete ===")

    # Exit with appropriate code
    sys.exit(0 if workflow_success else 1)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Plan-Implement workflow with Teamwork task updates"
    )
    parser.add_argument("adw_id", help="ADW execution ID")
    parser.add_argument("task_id", help="Teamwork task ID")
    parser.add_argument("task_description", help="Task description")
    parser.add_argument("worktree_name", help="Git worktree name")
    parser.add_argument("prototype", nargs="?", default="", help="Prototype type")
    parser.add_argument("--model", default="sonnet", help="Claude model")
    parser.add_argument("--project-id", help="Teamwork project ID")

    args = parser.parse_args()
    main(
        args.adw_id,
        args.task_id,
        args.task_description,
        args.worktree_name,
        args.prototype,
        args.model,
        args.project_id
    )
